lib/parsers/cr2_parser.rb

```diff
@@ -0,0 +1,119 @@
+class FormatParser::CR2Parser
+  include FormatParser::IOUtils
+
+  # Variables
+  TIFF_HEADER = [0x49, 0x49, 0x2a, 0x00]
+  CR2_HEADER  = [0x43, 0x52, 0x02, 0x00]
+
+  PREVIEW_ORIENTATION_TAG = 0x0112
+  PREVIEW_RESOLUTION_TAG = 0x011a
+  PREVIEW_IMAGE_OFFSET_TAG = 0x0111
+  PREVIEW_IMAGE_BYTE_COUNT_TAG = 0x0117
+
+  def call(io)
+    io = FormatParser::IOConstraint.new(io)
+
+    tiff_header = safe_read(io, 8)
+
+    # Check whether it's a CR2 file
+    tiff_bytes = tiff_header[0..3].bytes
+    magic_bytes = safe_read(io, 4).unpack('C4')
+
+    return if !magic_bytes.eql?(CR2_HEADER) || !tiff_bytes.eql?(TIFF_HEADER)
+
+    # Offset to IFD #0 where the preview image data is located
+    # For more information about CR2 format,
+    # see http://lclevy.free.fr/cr2/
+    # and https://github.com/lclevy/libcraw2/blob/master/docs/cr2_poster.pdf
+    if0_offset = tiff_header[4..7].reverse.bytes.collect { |c| c.to_s(16) }.join.hex
+
+    set_orientation(io, if0_offset)
+    set_resolution(io, if0_offset)
+    set_preview(io, if0_offset)
+
+    # Check CanonAFInfo or CanonAFInfo2 tags in maker notes for width & height
+    exif_offset = parse_ifd(io, if0_offset, 0x8769)
+    makernote_offset = parse_ifd(io, exif_offset[0], 0x927c)
+    af_info = parse_ifd(io, makernote_offset[0], 0x0026)
+
+    # Old Canon models have CanonAFInfo tags (0x0012)
+    # Newer models have CanonAFInfo2 tags (0x0026) instead
+    # See https://sno.phy.queensu.ca/~phil/exiftool/TagNames/Canon.html
+    if !af_info.nil?
+      parse_new_model(io, af_info[0], af_info[1])
+    else
+      af_info = parse_ifd(io, makernote_offset[0], 0x0012)
+      parse_old_model(io, af_info[0], af_info[1])
+    end
+
+    FormatParser::Image.new(
+      format: :cr2,
+      width_px: @width,
+      height_px: @height,
+      orientation: @orientation,
+      image_orientation: @image_orientation,
+      resolution: @resolution,
+      preview: parse_preview_image(io)
+    )
+  end
+
+  private
+
+  def parse_ifd(io, offset, searched_tag)
+    io.seek(offset)
+    entries_count = to_hex safe_read(io, 2)
+    entries_count.times do
+      entry = safe_read(io, 12)
+      tag_id = to_hex(entry[0..1])
+      length = to_hex(entry[4..7])
+      value = to_hex(entry[8..11])
+      return [value, length] if tag_id == searched_tag
+    end
+    nil
+  end
+
+  def to_hex(sequence)
```

Comment:
Is this method really doing `to_hex` (converting to hex)? It seems that what it actually does is a "bin to hex" conversion, but it's output is an integer - and it applies an integer conversion on top. It can also be rewritten to avoid string allocations that it does right now:

```ruby
[47] pry(main)> '100'.reverse.unpack('H*').join.hex
=> 3158065
[48] pry(main)> '100'.reverse.unpack('H*').join.hex
```

Doesn't seem we can ditch the `reverse` because apparently there is no flipped byte order hex specifier for `pack`. Regardless I would give the method a better name, what could it be?

Reply:
This is a much better approach. Thank you very much for that. I'l try to come up with a better name for ```to_hex```method.

Reply:
I have renamed```to_hex```method to ```parse_sequence_to_int```